### PR TITLE
Minor cleanups to Tillig's great work

### PR DESCRIPTION
**Pull request by nulltoken**
Follow up to #5 

**Checklist**

- [x] Tests added / updated
- [x] Docs added / updated

**Does this PR introduce a breaking change?**

- [x] Yes
- [ ] No

> If indicated yes above, please describe the breaking change(s).
>
> **Remove this quote before creating the PR.**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,5 +1,6 @@
-out/
-dev/
+artifacts/
 node_modules/
+out/
+wbpkd/
+.dist/
 .vscode-test/
```
Comment (P0lip) [inline .gitignore:6]: let's include `.idea` directory either, so JetBrains IDE users (like myself :wink:) are not harmed.

File: README.md [hunk 2]
```diff
@@ -20,39 +20,23 @@ Visual Studio Code version 1.37 or higher.
...
-
-## Known Issues / Limitations
-
-If you open a single file in VS Code (as opposed to opening a folder or
-workspace), then [Spectral rulesets](https://stoplight.io/p/docs/gh/stoplightio/spectral/docs/getting-started/rulesets.md) `(.?spectral.{yml|yaml|json}` files) will
-not be found, even if they are in the same directory. This is due to the use
-of the recommended/performant VS Code APIs for finding and watching files,
-which are exclusively workspace-based.
+* `spectral.enable`: Controls whether or not Spectral is enabled.
+* `spectral.rulesetFile`: Location of the ruleset file to use when validating. If omitted, the default is a `.spectral.yml`/`.spectral.json` in the same folder as the document being validated. Paths are relative to the workspace.
```
Comment (P0lip) [inline README.md:31]:
```suggestion
* `spectral.rulesetFile`: Location of the ruleset file to use when validating. If omitted, the default is a `.spectral.yaml/.spectral.yml/.spectral.json/spectral.yaml/spectral.yml/spectral.json` in the same folder as the document being validated. Paths are relative to the workspace.
```
:laughing:

File: server/src/server.ts [hunk 3]
```diff
@@ -0,0 +1,380 @@
...
+          connection.sendNotification(StartWatcherNotification.type, { path: rulesetFile });
+          try {
+            settings.ruleset = await readRuleset(rulesetFile);
+          } catch (err) {
+            showErrorMessage(docPath, `Unable to read ruleset at ${rulesetFile}.`);
+          }
+        } else {
+          // If there's no ruleset available, default to OAS rules.
+          // https://github.com/stoplightio/spectral/blob/4928b6155659cec225d0cfab6d77d849d691cc24/src/cli/services/linter/utils/getRuleset.ts#L22
+          settings.ruleset = await readRuleset('spectral:oas');
```
Comment (P0lip) [inline server/src/server.ts:212]: fwiw, we could also load `asyncapi`.

File: client/src/test/e2e/helper.ts [hunk 4]
```diff
@@ -0,0 +1,51 @@
+/* eslint-disable valid-jsdoc */
+/* eslint-disable require-jsdoc */
```
Comment (P0lip) [inline client/src/test/e2e/helper.ts:2]:
```suggestion
/* eslint-disable valid-jsdoc, require-jsdoc */
```

File: server/src/server.ts [hunk 5]
```diff
@@ -0,0 +1,380 @@
...
+            // Don't validate the ruleset with itself.
+            settings.validate = false;
+            return settings;
+          }
+
+          connection.sendNotification(StartWatcherNotification.type, { path: rulesetFile });
+          try {
+            settings.ruleset = await readRuleset(rulesetFile);
+          } catch (err) {
+            showErrorMessage(docPath, `Unable to read ruleset at ${rulesetFile}.`);
```
Comment (P0lip) [inline server/src/server.ts:207]: This is likely to be an extra feature, but would be worth to validate the Spectral ruleset itself.
The error message saying we were unable to load a file won't be particularly helpful.
You can leverage `assertValidRuleset` function exposed by Spectral.

File: server/src/server.ts [hunk 6]
```diff
@@ -0,0 +1,380 @@
...
+      // Send diagnostics with the results to the client. If there aren't any
+      // results, it means the document was ignored or is valid - send an empty
+      // array of results to ensure any existing issues are cleared.
+      const diagnostics: Diagnostic[] = [];
+      if (results.length > 0) {
+        results.forEach((result) => {
+          const diagnostic = makeDiagnostic(result);
+          diagnostics.push(diagnostic);
+        });
+      }
```
Comment (P0lip) [inline server/src/server.ts:299]:
```suggestion
      const diagnostics = results.map<Diagnostic>(makeDiagnostic);
```
How about this?

File: server/src/util.ts [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+/**
+ * Converts a Spectral rule violation to a VS Code diagnostic.
+ * @param {IRuleResult} problem - The Spectral rule result to convert to a VS Code diagnostic message.
+ * @return {IDiagnostic} The converted VS Code diagnostic to send to the client.
+ */
+export function makeDiagnostic(problem: IRuleResult): Diagnostic {
+  return {
+    range: {
+      start: {
+        line: problem.range.start.line,
```
Comment (P0lip) [inline server/src/util.ts:37]: Does LSP expect 1-based or 0-based lines & columns?
Don't recall to be honest. I know VS Code's Monaco demands 1-based values.

File: README.md (hunk 2 above)
Comment (philsturgeon) [reply README.md:31]: Let's not mention the spectral.yaml but. I feel like that's a bug not a feature?

File: server/src/util.ts (hunk 7 above)
Comment (nulltoken) [reply server/src/util.ts:41]: @P0lip Those work as-is. VSCode seems to properly highlight the proper ranges. Thnx for having raised this!

File: server/src/server.ts (hunk 6 above)
Comment (nulltoken) [reply server/src/server.ts]: @P0lip It's been fixed differently.

File: server/src/server.ts (hunk 5 above)
Comment (nulltoken) [reply server/src/server.ts]: @philsturgeon Could you please keep track of that in your roadmap document?

File: client/src/test/e2e/helper.ts (hunk 4 above)
Comment (nulltoken) [reply client/src/test/e2e/helper.ts]: @P0lip Fixed

File: server/src/server.ts (hunk 3 above)
Comment (nulltoken) [reply server/src/server.ts]: @P0lip Fixed

File: README.md (hunk 2 above)
Comment (nulltoken) [reply README.md:31]: @philsturgeon Huh? I'm sorry but I didn't get this. What rewording would you like to see?

File: .gitignore (hunk 1 above)
Comment (nulltoken) [reply .gitignore:7]: @P0lip Fixed

**Pull request merged**